File: algorithms/0024.swap-nodes-in-pairs/solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def swap_pairs(head)
  if head&.next
    tmp = head.next.next
    head, head.next.next = head.next, head
    head.next.next = swap_pairs(tmp)
  end

  head
end

File: algorithms/0024.swap-nodes-in-pairs/solution.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	if head != nil && head.Next != nil {
		tmp := head.Next.Next
		head, head.Next.Next = head.Next, head
		head.Next.Next = swapPairs(tmp)
	}

	return head
}
